cmd/aws-cost-exporter/main.go

```diff
@@ -1,44 +1,74 @@
 package main
 
 import (
-	"github.com/st8ed/aws-cost-exporter/pkg/collector"
-	"github.com/st8ed/aws-cost-exporter/pkg/fetcher"
-	"github.com/st8ed/aws-cost-exporter/pkg/processor"
-	"github.com/st8ed/aws-cost-exporter/pkg/state"
-
 	"context"
+	"net/http"
+	"os"
 	"os/user"
+	"sync/atomic"
+	"time"
 
 	"github.com/aws/aws-sdk-go-v2/config"
 	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/collectors"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 	dto "github.com/prometheus/client_model/go"
-	"github.com/prometheus/common/version"
-
-	"github.com/go-kit/log"
-	"github.com/go-kit/log/level"
 	"github.com/prometheus/common/promlog"
 	"github.com/prometheus/common/promlog/flag"
-
-	"net/http"
-	"os"
-
+	"github.com/prometheus/common/version"
 	"github.com/prometheus/exporter-toolkit/web"
-
+	"golang.org/x/sync/errgroup"
 	kingpin "gopkg.in/alecthomas/kingpin.v2"
+
+	"github.com/st8ed/aws-cost-exporter/pkg/collector"
+	"github.com/st8ed/aws-cost-exporter/pkg/fetcher"
+	"github.com/st8ed/aws-cost-exporter/pkg/processor"
+	"github.com/st8ed/aws-cost-exporter/pkg/state"
 )
 
-func newGatherer(config *state.Config, state *state.State, client *s3.Client, disableExporterMetrics bool, logger log.Logger) (prometheus.GathererFunc, error) {
-	reg := prometheus.NewRegistry()
+type intervalGatherer struct {
+	logger   log.Logger
+	gatherer prometheus.Gatherer
+	cache    atomic.Pointer[[]*dto.MetricFamily]
+}
 
-	if !disableExporterMetrics {
-		reg.MustRegister(collectors.NewBuildInfoCollector())
-		reg.MustRegister(collectors.NewGoCollector(
-			collectors.WithGoCollections(collectors.GoRuntimeMemStatsCollection | collectors.GoRuntimeMetricsCollection),
-		))
+// Gather implements prometheus.Gatherer.
+func (ig *intervalGatherer) Gather() ([]*dto.MetricFamily, error) {
+	mfs := ig.cache.Load()
+	if mfs == nil {
+		return nil, nil
+	}
+	return *mfs, nil
+}
+
+// Run will gather metrics from the gatherer at the given interval until the context is done.
+func (ig *intervalGatherer) Run(ctx context.Context, interval time.Duration) error {
+	// Gather once on startup.
+	mfs, err := ig.gatherer.Gather()
+	if err != nil {
+		return err
 	}
+	ig.cache.Store(&mfs)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(interval):
+			mfs, err := ig.gatherer.Gather()
+			if err != nil {
+				level.Warn(ig.logger).Log("err", err)
+				continue
```

Comment:
I think it could be better to `return err` here too (as in the first `.Gather()` call), so the exporter closes immediately with an error code.

Otherwise we may be exporting stale metrics indefinitely, and that's something that I would very much like to prevent. 

Or we could consider adding exporting the time of last computed report and set alarms if it is too far, but it's more difficult to setup.

Reply:
hmm that's a good point! let's error then.